blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/util/UserSSHKeyManager.java

```diff
@@ -0,0 +1,214 @@
+        
+        try {
+            // create one!
+            CredentialsStore store = getUserStore(user);
+            if(store == null){
+                throw new ServiceException.ForbiddenException(String.format("Logged in user: %s doesn't have writable credentials store", user.getId()));
+            }
+            
+            Credentials key = null;
+            // try to find the key
+            for (Credentials cred : store.getCredentials(getDomain(store))) {
+                if (cred instanceof BasicSSHUserPrivateKey) {
+                    BasicSSHUserPrivateKey sshKey = (BasicSSHUserPrivateKey)cred;
+                    if (BLUEOCEAN_GENERATED_SSH_KEY_ID.equals(sshKey.getId())) {
+                        key = sshKey;
+                        break;
+                    }
+                }
+            }
+            if (key != null) {
+                store.removeCredentials(getDomain(store), key);
+                store.save();
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    /**
+     * Gets the user's CredentialStore
+     * @param user user to find a store for
+     * @return the credential store or null if not found
+     */
+    private static @CheckForNull CredentialsStore getUserStore(User user){
+        for(CredentialsStore s : CredentialsProvider.lookupStores(user)) {
+            if(s.hasPermission(CredentialsProvider.CREATE) && s.hasPermission(CredentialsProvider.UPDATE)){
+                return s;
+            }
+        }
+        return null;
+    }
+    
+    /**
+     * Gets an "ssh-rsa"-style formatted public key with useful identifier
+     */
+    private static String getSSHPublicKey(RSAPublicKey key, User user) throws IOException {
+        return "ssh-rsa " + Base64.encode(SSHKeyUtils.encodePublicKey(key)) + " " + getKeyComment(user.getId());
+    }
+
+    /**
+     * Returns a <user>@<jenkins-host> sort of comment to help identify the key's origin
+     * @param userId a user id
+     * @return an identifier
+     */
+    private static String getKeyComment(String userId) {
+        String host = Jenkins.getInstance().getRootUrl();
+        if (host == null) {
+            host = Jenkins.getInstance().getRootUrlFromRequest();
+        }
+        host = host.replaceAll(".*//([^/]+).*", "$1");
+        return ((userId == null ? Jenkins.getInstance().getDisplayName() : userId) + "@" + host)
+                .replaceAll("[^:@._a-zA-Z0-9]", "");
+    }
+    
+    private static Domain getDomain(CredentialsStore store) {
+        Domain domain = store.getDomainByName(BLUEOCEAN_DOMAIN_NAME);
+        if (domain == null) {
+            try {
+                //create new one
+                boolean result = store.addDomain(new Domain(BLUEOCEAN_DOMAIN_NAME, null, null));
+                if (!result) {
+                    throw new RuntimeException(String.format("Failed to create credential domain: %s", BLUEOCEAN_DOMAIN_NAME));
```

Comment:
better to throw `UnexpectedErrorException`. It gets proper json response rendering.